File: forms/pre-operative-assessment-by-patient/front-end-patient-form-with-html/js/asa-grader.js
// ──────────────────────────────────────────────
// ASA grading engine
// ──────────────────────────────────────────────

import { asaRules } from './asa-rules.js';

/**
 * Pure function: evaluates all ASA rules against patient data.
 * Returns the maximum grade among all fired rules (worst comorbidity),
 * defaulting to ASA I for healthy patients with no fired rules.
 */
export function calculateASA(data) {
	const firedRules = [];

	for (const rule of asaRules) {
		try {
			if (rule.evaluate(data)) {
				firedRules.push({
					id: rule.id,
					system: rule.system,
					description: rule.description,
					grade: rule.grade
				});
			}
		} catch (e) {
			console.warn(`ASA rule ${rule.id} evaluation failed:`, e);
		}
	}

	const asaGrade =
		firedRules.length === 0
			? 1
			: Math.max(...firedRules.map((r) => r.grade));

	return { asaGrade, firedRules };
}

File: forms/pre-operative-assessment-by-patient/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/asa_grader.rs
use super::asa_rules::all_rules;
use super::types::{AsaGrade, AssessmentData, FiredRule};

/// Pure function: evaluates all ASA rules against patient data.
/// Returns the maximum grade among all fired rules (worst comorbidity),
/// defaulting to ASA I for healthy patients with no fired rules.
pub fn calculate_asa(data: &AssessmentData) -> (AsaGrade, Vec<FiredRule>) {
    let mut fired_rules = Vec::new();

    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                system: rule.system.to_string(),
                description: rule.description.to_string(),
                grade: rule.grade,
            });
        }
    }

    let asa_grade = if fired_rules.is_empty() {
        1
    } else {
        fired_rules.iter().map(|r| r.grade).max().unwrap_or(1)
    };

    (asa_grade, fired_rules)
}
